src/Deserialization.cpp

```diff
@@ -0,0 +1,1358 @@
+    case Serialize::SplitType::SplitType_RenameVar:
+        return Split::SplitType::RenameVar;
+    case Serialize::SplitType::SplitType_FuseVars:
+        return Split::SplitType::FuseVars;
+    case Serialize::SplitType::SplitType_PurifyRVar:
+        return Split::SplitType::PurifyRVar;
+    default:
+        user_assert(false) << "unknown split type " << split_type << "\n";
+        return Split::SplitType::SplitVar;
+    }
+}
+
+DimType Deserializer::deserialize_dim_type(const Serialize::DimType dim_type) {
+    switch (dim_type) {
+    case Serialize::DimType::DimType_PureVar:
+        return DimType::PureVar;
+    case Serialize::DimType::DimType_PureRVar:
+        return DimType::PureRVar;
+    case Serialize::DimType::DimType_ImpureRVar:
+        return DimType::ImpureRVar;
+    default:
+        user_assert(false) << "unknown dim type " << dim_type << "\n";
+        return DimType::PureVar;
+    }
+}
+
+LoopAlignStrategy Deserializer::deserialize_loop_align_strategy(const Serialize::LoopAlignStrategy loop_align_strategy) {
+    switch (loop_align_strategy) {
+    case Serialize::LoopAlignStrategy::LoopAlignStrategy_AlignStart:
+        return LoopAlignStrategy::AlignStart;
+    case Serialize::LoopAlignStrategy::LoopAlignStrategy_AlignEnd:
+        return LoopAlignStrategy::AlignEnd;
+    case Serialize::LoopAlignStrategy::LoopAlignStrategy_NoAlign:
+        return LoopAlignStrategy::NoAlign;
+    case Serialize::LoopAlignStrategy::LoopAlignStrategy_Auto:
+        return LoopAlignStrategy::Auto;
+    default:
+        user_assert(false) << "unknown loop align strategy " << loop_align_strategy << "\n";
+        return LoopAlignStrategy::AlignStart;
+    }
+}
+
+ExternFuncArgument::ArgType Deserializer::deserialize_extern_func_argument_type(const Serialize::ExternFuncArgumentType extern_func_argument_type) {
+    switch (extern_func_argument_type) {
+    case Serialize::ExternFuncArgumentType::ExternFuncArgumentType_UndefinedArg:
+        return ExternFuncArgument::ArgType::UndefinedArg;
+    case Serialize::ExternFuncArgumentType::ExternFuncArgumentType_FuncArg:
+        return ExternFuncArgument::ArgType::FuncArg;
+    case Serialize::ExternFuncArgumentType::ExternFuncArgumentType_BufferArg:
+        return ExternFuncArgument::ArgType::BufferArg;
+    case Serialize::ExternFuncArgumentType::ExternFuncArgumentType_ExprArg:
+        return ExternFuncArgument::ArgType::ExprArg;
+    case Serialize::ExternFuncArgumentType::ExternFuncArgumentType_ImageParamArg:
+        return ExternFuncArgument::ArgType::ImageParamArg;
+    default:
+        user_assert(false) << "unknown extern func argument type " << extern_func_argument_type << "\n";
+        return ExternFuncArgument::ArgType::UndefinedArg;
+    }
+}
+
+Type Deserializer::deserialize_type(const Serialize::Type *type) {
+    user_assert(type != nullptr) << "deserializing a null Type\n";
+    using Serialize::TypeCode;
+    int bits = type->bits();
```

Comment:
Style nit, here and a lot of other places: it's (slightly) preferable code style to define locals like this as `const` when you know they won't/can't be mutated later on; it makes the code a bit easier to read when you can immediately note that this local won't change later based on something you didn't expect.